waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -49,6 +58,49 @@ validateExactlyOneAppExists spec =
         GenericValidationError $
           "You have more than one 'app' declaration in your Wasp app. You have " ++ show (length apps) ++ "."
 
+validateWasp :: AppSpec -> [ValidationError]
+validateWasp spec = validateWaspVersion specWaspVersionStr
+  where
+    specWaspVersionStr = Wasp.version $ App.wasp (snd $ getApp spec)
+
+validateWaspVersion :: String -> [ValidationError]
+validateWaspVersion specWaspVersionStr = eitherUnitToErrorList $ do
+  specWaspVersionRange <- parseWaspVersionRange specWaspVersionStr
+  unless
+    ( SV.isVersionInRange currentWaspVersion specWaspVersionRange
+    )
+    $ Left $ incompatibleVersionError currentWaspVersion specWaspVersionRange
+  where
+    parseWaspVersionRange :: String -> Either ValidationError SV.Range
+    parseWaspVersionRange waspVersionRangeStr = do
+      let (_ :: String, _ :: String, _ :: String, waspVersionRangeDigits :: [String]) =
+            waspVersionRangeStr =~ ("\\^([0-9]+).([0-9]+).([0-9]+)$" :: String)
+
+      waspSpecVersion <- case mapM readMaybe waspVersionRangeDigits of
+        Just [major, minor, patch] -> Right $ SV.Version major minor patch
+        __ -> Left $ GenericValidationError "Wasp version should be in the format ^major.minor.patch"
+
+      Right $ SV.Range [SV.backwardsCompatibleWith waspSpecVersion]
+
+    incompatibleVersionError :: SV.Version -> SV.Range -> ValidationError
+    incompatibleVersionError actualVersion expectedVersionRange =
+      GenericValidationError $
+        unwords
+          [ "Your Wasp version does not match the app's requirements.",
+            "You are running Wasp " ++ show actualVersion ++ ".",
+            "This app requires Wasp " ++ show expectedVersionRange ++ "."
+          ]
+
+    currentWaspVersion :: SV.Version
+    currentWaspVersion = SV.Version (intToNat major) (intToNat minor) (intToNat patch)
+      where
+        DV.Version [major, minor, patch] _ = Paths_waspc.version
+        intToNat = naturalFromInteger . toInteger
```

Comment:
You can use the already available [`toEnum` function](https://hoogle.haskell.org/?hoogle=Int+-%3E+Natural&scope=set%3Ahaskell-platform) instead of defining your own.

Reply:
This one also comes from @Martinsos 's review. In fact, I was using `toEnum` in the first iteration of this PR:

> Finally, I wasn't sure what happens when you call toEnum to convert Int to Natural, so I implemented function for it instead.

Reply:
Whoop, :sweat_smile: 

My bad, missed that comment. I thought there was a chance it might have been done for readability purposes. It's fine either way.

@Martinsos, can't say I fully understand the reasoning though, did you want to say "_it's easier for someone unfamiliar with the API to anticipate what to expect when calling `toInteger` and `naturalFromInteger` than `toEnum` (i.e., this last one might return something weird)?_"

Reply:
Yup @NeoLight1010 actually used `toEnum` but I suggested this!
To be honest I was going pretty quickly through this and haven't properly investigated what `toEnum` is but concluded that it sounds suspicious and went with this hand rolled solution. Maybe I should have investigated it more though. The thing is, it converts Int to Enum. But Enum is just something enumerated, it doesn't know it is a number. So I felt like there are no guarantees here that it will do reasonable number to number conversion -> Enum cares just about order, not about absolute values.

Looking at it now, I still don't know if toEnum guarantees me what I would expect in this case hm!

On the other hand:
 - `toInteger` -> yup, I do expect this one to convert from Int to corresponding Integer.
 - `naturalFromInteger` -> also, this one is very targeted, I am sure it will give me corresponding natural, and throw an error or something if Integer is not natural.

So I am not sure how to go about this, it does seem a bit silly to write this when there is `toEnum`, and it works as we need it I believe, but in theory I am not sure what is  `toEnum` promising is enough, and I do think reading it in code is a bit confusing (why I am converting number to enum?).

Summary: I am good with using `toEnum` if you like, but it is a bit weird function I think, to be used like this! But maybe I am overthinking it.

Reply:
I'd say it's just a standard library function with a bad name. Every language has its fair share of those. I don't think we should be implementing our own version of the same function because we don't like its name - it is what it is :)

Think about PHP's `implode` (a string join operation). It's a horrible name (IMO). Still, if you're a PHP developer, you're stuck with it. Writing your own function that does the same thing and naming it `join` or `intercalate` (also a weird name btw) will only raise questions with readers familiar with the language: "Why did they write their own `implode`? is this an edge case in which they differ? Is the original function not performant enough? Does this one even work correctly?..." 
Finally, they'll maybe have to read (or at least glance at) the source code of your function to figure out what it does. This is not the case when dealing with a standard function (regardless of how bad its name is).

With that said, in this case it's a pretty small and irrelevant change, so we can go either way as far as I'm concerned. But if I'm following my gut, I'd personally just choose `toEnum`.

Reply:
I agree with @sodic here. Also, wow, `implode` is such a bad name o.0 .

I can go with `toEnum`, if that's ok :smile: